### PR TITLE
Add transaction.context

### PR DESCRIPTION
**Pull request by mikker**

### CODE REVIEW

File: lib/elastic_apm/context.rb [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        @headers = {}
+        @cookies = {}
+        @env = {}
+      end
+
+      attr_accessor :body, :cookies, :env, :headers, :http_version, :method,
+        :socket, :url
+
+      # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+      def add_rack_env(rack_env)
```
Comment (simitt) [inline lib/elastic_apm/context.rb]: I find the handling in `add_rack_env` relatively confusing as I cannot find a consistent usage of virtual attrs, initializing hashes in the init method, accessing and setting attrs, etc. 

If `add_rack_env` would be called twice on the same object, then the information in `socket` would be set according to the second `req`, but `headers` would be basically merged from the first and the second `req`. This is an outcome of once overwriting `socket` completely, but only adding key-values to `headers`, while both initializing in the init method. 
I know this is theoretical, but it shows how different you are initializing and accessing the attrs here. 

Maybe you find a way to make this more consistent.

Comment (mikker) [reply lib/elastic_apm/context.rb]: Thanks @simitt. You are completely right that this is a bit confusing. I'll try and refactor it to make more sense.

**Comment by mikker** [comment]
@simitt While this in some ways just moves things around, I think it also might make its behaviour less weird -- _theoretically™_ 😀 

The distance between the specs testing this and the functionality might be too long now though. Will have to move things around there as well to keep everyone sane in the long run.

File: lib/elastic_apm/context.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+module ElasticAPM
+  # @api private
+  class Context
+    # @api private
+    class Request
+      def initialize
+        @socket = {}
+        @headers = {}
+        @cookies = {}
+        @env = {}
```
Comment (simitt) [inline lib/elastic_apm/context.rb]: I think you are missing `url` here.

Comment (mikker) [reply lib/elastic_apm/context.rb]: I don't think there's any reason to initialize these anymore.

**Pull request merged**